integrations/snyk/snyk/client.py

```diff
@@ -0,0 +1,268 @@
+from typing import Any, Optional, AsyncGenerator
+import httpx
+from loguru import logger
+
+from port_ocean.context.event import event
+
+
+class SnykClient:
+    def __init__(
+        self,
+        token: str,
+        api_url: str,
+        app_host: str | None,
+        org_id: str,
+        webhook_secret: str | None,
+    ):
+        self.token = token
+        self.api_url = f"{api_url}/v1"
+        self.app_host = app_host
+        self.org_id = org_id
+        self.rest_api_url = f"{api_url}/rest"
+        self.webhook_secret = webhook_secret
+        self.http_client = httpx.AsyncClient(headers=self.api_auth_header, timeout=30)
+        self.snyk_api_version = "2023-08-21"
+
+    @property
+    def api_auth_header(self) -> dict[str, Any]:
+        return {"Authorization": f"token {self.token}"}
+
+    async def _send_api_request(
+        self,
+        url: str,
+        method: str = "GET",
+        query_params: Optional[dict[str, Any]] = None,
+        json_data: Optional[dict[str, Any]] = None,
+    ) -> dict[str, Any]:
+        try:
+            response = await self.http_client.request(
+                method=method, url=url, params=query_params, json=json_data
+            )
+            response.raise_for_status()
+            return response.json()
+
+        except httpx.HTTPStatusError as e:
+            logger.error(
+                f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+            )
+            raise
+
+    async def _get_paginated_resources(
+        self,
+        base_url: str,
+        url_path: str,
+        method: str = "GET",
+        data_key: str = "data",
+        query_params: Optional[dict[str, Any]] = None,
+    ) -> AsyncGenerator[list[Any], None]:
+        while url_path:
+            try:
+                full_url = f"{base_url}{url_path}"
+
+                response = await self.http_client.request(
+                    method=method,
+                    url=full_url,
+                    params={**(query_params or {}), "limit": 50},
+                )
+                response.raise_for_status()
+                data = response.json()
+
+                yield data[data_key]
+
+                # Check if there is a "next" URL in the links object
+                url_path = data.get("links", {}).get("next")
+            except httpx.HTTPStatusError as e:
+                logger.error(
+                    f"HTTP error with status code: {e.response.status_code} and response text: {e.response.text}"
+                )
+                raise
+
+    async def get_issues(self, project: dict[str, Any]) -> list[dict[str, Any]]:
+        def build_cache_key(project_id: str) -> str:
+            return f"issues-{project_id}"
+
```

Comment:
I would add a comment that we cache the issues for the resync mechanism as there are multiple resync kinds that needs the issues per project and the resync has the same event context for all resync methods